### PR TITLE
Show a disabled geolocate button if user or system denies geolocation

### PR DESCRIPTION
**Pull request by kkaefer**
Fixes https://github.com/mapbox/mapbox-gl-js/issues/8870

image

This PR adds a new icon to be shown when the user or the system denied permission to use the location. It's shown when geolocation is disabled outright in Firefox and Chrome, or when the user or the system explicitly denies location after being prompted.

There are a few caveats:
- It's not possible to detect whether the user has denied location to begin with when loading a page. Safari doesn't implement [navigator.permissions](https://developer.mozilla.org/en-US/docs/Web/API/Navigator/permissions) detection. Once the user has click on the geolocate button, the state is set correctly though.
- When macOS location services are disabled _on a system-wide level_, Firefox never calls any of the callbacks, and the location indicator keeps spinning. This is a bug in Firefox and tracked in https://bugzilla.mozilla.org/show_bug.cgi?id=1529591
- When macOS location services are disabled, Chrome sneakily circumvents the OS setting and generates a location from different sources (e.g. IP address or Wifi networks; in my case, the location was approximate, but still in the same neighborhood).

Before this patch, the geolocate button was hidden when geolocation was disabled. It is now being shown, but in the new disabled state.

I'm a bit unsure how to test this correctly, since it depends on complex interaction between the OS, and the user clicking on various browser UIs.

/cc @andrewharvey

### CODE REVIEW

File: src/ui/control/geolocate_control.js [hunk 1]
```diff
@@ -287,17 +288,17 @@ class GeolocateControl extends Evented {
...
-        }
         this._container.addEventListener('contextmenu', (e: MouseEvent) => e.preventDefault());
         this._geolocateButton = DOM.create('button',
             `${className}-icon ${className}-geolocate`,
             this._container);
         this._geolocateButton.type = 'button';
         this._geolocateButton.title = 'Find my location';
         this._geolocateButton.setAttribute('aria-label', 'Find my location');
+        if (supported === false) {
+            warnOnce('Geolocation support is not available, the GeolocateControl will be disabled.');
```
Comment (ryanhamley) [inline src/ui/control/geolocate_control.js]: Nit: I know this wasn't introduced here, but it bugs me. Can we either make this two sentences or make it `so the GeolocateControl...`?

**Comment by ryanhamley** [comment]
>I'm a bit unsure how to test this correctly, since it depends on complex interaction between the OS, and the user clicking on various browser UIs.

Another argument for adding an end-to-end test suite

**Comment by andrewharvey** [comment]
Could we update the title/aria-label when disabled to be something like "Location not available"?

**Comment by andrewharvey** [comment]
This is good!

We should make sure to include a note in the changelog since the behaviour has changed. Anyone who prefers to hide the control when not available can still override the css `.mapboxgl-ctrl-geolocate:disabled { display: none }`.

**Comment by kkaefer** [comment]
I tested in Safari in the following way:

```bash
brew install mkcert
mkcert -install
mkcert localhost

python -c "import BaseHTTPServer, SimpleHTTPServer, ssl; \
           handler = SimpleHTTPServer.SimpleHTTPRequestHandler; \
           httpd = BaseHTTPServer.HTTPServer(('localhost', 4443), handler); \
           httpd.socket = ssl.wrap_socket(httpd.socket, \
                                          certfile='./localhost.pem', \
                                          keyfile='./localhost-key.pem', \
                                          server_side=True); \
           httpd.serve_forever()"
```

Then accessed the server at `https://localhost:4443/debug/debug.html`

**Pull request merged**